patient-appointment-management/functions/auth.private.js

```diff
@@ -1,19 +1,77 @@
-/* eslint-disable no-negated-condition, no-else-return */
+/* eslint-disable no-negated-condition, no-else-return, import/newline-after-import, no-use-before-define, vars-on-top, no-var, object-shorthand, sonarjs/prefer-immediate-return */
 const crypto = require('crypto');
+var jwt = require('jsonwebtoken');
+const MFA_TOKEN_DURATION = 5 * 60;
+const APP_TOKEN_DURATION = 30 * 60;
 
-function createToken(password, context) {
-  const tokenString = `${context.ACCOUNT_SID}:${password}:${context.SALT}`;
+function isValidPassword(password, context) {
+  return (
+    checkDisableAuthForLocalhost(context) ||
+    password === context.APPLICATION_PASSWORD
+  );
+}
+
+function checkDisableAuthForLocalhost(context) {
+  return (
+    context.DOMAIN_NAME &&
+    context.DOMAIN_NAME.startsWith('localhost') &&
+    context.DISABLE_AUTH_FOR_LOCALHOST &&
+    context.DISABLE_AUTH_FOR_LOCALHOST === 'true'
+  );
+}
 
-  return crypto
-    .createHmac('sha1', context.AUTH_TOKEN)
-    .update(Buffer.from(tokenString, 'utf-8'))
+function createMfaToken(issuer, mfaCode, context) {
+  if (checkDisableAuthForLocalhost(context)) {
+    return createAppToken(issuer, context);
+  }
+
+  // encrypt the mfaCode to avoid showing in the browser
+  mfaEncrypt = crypto
+    .createHmac('sha256', context.AUTH_TOKEN)
+    .update(Buffer.from(`${mfaCode}:${context.SALT}`, 'utf-8'))
     .digest('base64');
+
+  var jwtToken = jwt.sign({ data: mfaEncrypt }, context.AUTH_TOKEN, {
+    expiresIn: MFA_TOKEN_DURATION,
+    audience: 'mfa',
+    issuer: issuer,
+    subject: 'administrator',
+  });
+
+  return jwtToken;
 }
 
-function isAllowed(token, context) {
-  // Create the token with the environment password
-  const masterToken = createToken(context.APPLICATION_PASSWORD, context);
-  return masterToken === token;
+function isValidMfaToken(token, context) {
```

Comment:
with Verify this function becomes:

```
twilioClient.verify.services('VAXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXXX')
      .verificationChecks
      .create({to: context. ADMINISTRATOR_PHONE, code: token})
      .then(check => {
  return check.status === 'approved';
}));
```

And you can remove the jwt/mfaEncrypt/token timeout (Verify automatically handles expiration).

Let me know if I can help further!

Reply:
thanks for the detailed info. we'll convert to Twilio verify in v3. https://issues.corp.twilio.com/browse/HD-27